File: set_bits_debugging_judge.py
#!/usr/bin/env python3
"""
FixIt Platform - Set Bits Counting Problem Integration
Multi-language debugging problem with line comparison evaluation
"""

import json
import time
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum

class EvaluationStatus(Enum):
    PASSED = "PASSED"
    FAILED = "FAILED"
    CHEATING_DETECTED = "CHEATING_DETECTED"
    SYSTEM_ERROR = "SYSTEM_ERROR"

@dataclass
class BugDefinition:
    """Defines a single bug in a debugging problem"""
    bug_id: str
    buggy_line: str
    expected_fix: str
    description: str
    points: int = 10

@dataclass
class LanguageCode:
    """Code definitions for different languages"""
    buggy_code: str
    correct_code: str

@dataclass
class ProblemDefinition:
    """Complete debugging problem definition"""
    problem_id: int
    title: str
    description: str
    bugs: List[BugDefinition]
    total_points: int
    languages: Dict[str, LanguageCode]

class SetBitsDebuggingJudge:
    """Specialized judge for set bits counting debugging problem"""
    
    def __init__(self):
        self.problem = self.create_set_bits_problem()
    
    def create_set_bits_problem(self) -> ProblemDefinition:
        """Create the set bits counting debugging problem"""
        
        # Bug definitions (common across all languages)
        bugs = [
            BugDefinition(
                bug_id="count_initialization",
                buggy_line="count = 2",
                expected_fix="count = 0",
                description="Count should start from 0",
                points=10
            ),
            BugDefinition(
                bug_id="loop_condition",
                buggy_line="while n < 0",
                expected_fix="while n > 0",
                description="Loop should run while n is positive",
                points=15
            ),
            BugDefinition(
                bug_id="bitwise_operator",
                buggy_line="if (n | 1)",
                expected_fix="if (n & 1)",
                description="Use AND to check the last bit",
                points=15
            ),
            BugDefinition(
                bug_id="comparison_operator",
                buggy_line="= 1",
                expected_fix="== 1",
                description="Use comparison operator instead of assignment",
                points=15
            ),
            BugDefinition(
                bug_id="shift_direction",
                buggy_line="n = n << 1",
                expected_fix="n = n >> 1",
                description="Shift right to remove processed bit",
                points=15
            ),
            BugDefinition(
                bug_id="print_typo",
                buggy_line="print(cout)",
                expected_fix="print(count)",
                description="Variable name typo",
                points=10
            )
        ]
        
        # Language-specific code
        languages = {
            "python": LanguageCode(
                buggy_code='''n = int(input())

count = 2

while n < 0:
    if (n | 1) = 1:
        count = count + 1
    n = n << 1

print(cout)''',
                correct_code='''n = int(input())

count = 0

while n > 0:
    if (n & 1) == 1:
        count = count + 1
    n = n >> 1

print(count)'''
            ),
            "java": LanguageCode(
                buggy_code='''import java.util.*;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        int count = 2;

        while (n < 0) {
            if ((n | 1) = 1) {
                count = count + 1;
            }
            n = n << 1;
        }

        System.out.println(cout);
    }
}''',
                correct_code='''import java.util.*;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        int count = 0;

        while (n > 0) {
            if ((n & 1) == 1) {
                count = count + 1;
            }
            n = n >> 1;
        }

        System.out.println(count);
    }
}'''
            ),
            "cpp": LanguageCode(
                buggy_code='''#include <iostream>
using namespace std;

int main() {

    int n;
    cin >> n;

    int count = 2;

    while (n < 0) {
        if ((n | 1) = 1) {
            count = count + 1;
        }
        n = n << 1;
    }

    cout << cout;

    return 0;
}''',
                correct_code='''#include <iostream>
using namespace std;

int main() {

    int n;
    cin >> n;

    int count = 0;

    while (n > 0) {
        if ((n & 1) == 1) {
            count = count + 1;
        }
        n = n >> 1;
    }

    cout << count;

    return 0;
}'''
            )
        }
        
        return ProblemDefinition(
            problem_id=1,
            title="Count Set Bits - Debug the Algorithm",
            description="Given an integer N, count the number of set bits (1s) in its binary representation. Fix all bugs in the provided code.",
            bugs=bugs,
            total_points=80,
            languages=languages
        )
    
    def normalize_code(self, code: str, language: str) -> str:
        """Normalize code for consistent comparison"""
        if not code:
            return ""
        
        # Convert to string and strip whitespace
        code = str(code).strip()
        
        # Normalize line endings
        code = code.replace('\r\n', '\n').replace('\r', '\n')
        
        # Split into lines and normalize each line
        lines = code.split('\n')
        normalized_lines = []
        
        for line in lines:
            # Remove leading/trailing whitespace
            line = line.strip()
            
            # Normalize internal whitespace (multiple spaces -> single space)
            line = ' '.join(line.split())
            
            # Language-specific comment removal
            if language == "python":
                line = line.split('#')[0]  # Remove Python comments
            elif language in ["java", "cpp"]:
                line = line.split('//')[0]  # Remove C++/Java comments
                # Remove C-style comments
                if '/*' in line:
                    line = line.split('/*')[0]
            
            # Remove trailing whitespace again
            line = line.strip()
            
            # Only add non-empty lines
            if line:
                normalized_lines.append(line)
        
        return '\n'.join(normalized_lines)
    
    def normalize_line(self, line: str) -> str:
        """Normalize a single line for comparison"""
        if not line:
            return ""
        
        # Convert to string and strip whitespace
        line = str(line).strip()
        
        # Normalize internal whitespace
        line = ' '.join(line.split())
        
        return line
    
    def detect_cheating(self, code: str, language: str) -> Tuple[bool, List[str]]:
        """Detect potential cheating attempts"""
        violations = []
        normalized_code = self.normalize_code(code, language)
        
        # Check for hardcoded outputs
        hardcoded_patterns = [
            r'print\s*\(\s*\d+\s*\)',           # print(42)
            r'cout\s*<<\s*\d+',                  # cout << 42
            r'System\.out\.println\s*\(\s*\d+\s*\)',  # System.out.println(42)
            r'return\s+\d+',                       # return 42
        ]
        
        # Only check for obvious hardcoded patterns
        import re
        for pattern in hardcoded_patterns:
            matches = re.findall(pattern, normalized_code, re.IGNORECASE)
            if matches:
                violations.append(f"Hardcoded output detected: {pattern}")
        
        # Check for suspicious comments (only if they contain actual code)
        lines = normalized_code.split('\n')
        for i, line in enumerate(lines):
            # Only flag if line has actual comment patterns with code keywords
            if (('#' in line and any(keyword in line.lower() for keyword in ['if', 'while', 'for', 'return', 'print'])) or
                ('//' in line and any(keyword in line.lower() for keyword in ['if', 'while', 'for', 'return', 'print']))):
                violations.append(f"Suspicious commented code at line {i+1}: {line}")
        
        return len(violations) > 0, violations
    
    def evaluate_single_bug(self, normalized_code: str, bug: BugDefinition, language: str) -> Dict:
        """Evaluate if a single bug is fixed correctly"""
        # Normalize both buggy line and expected fix
        normalized_buggy = self.normalize_line(bug.buggy_line)
        normalized_fix = self.normalize_line(bug.expected_fix)
        
        # Language-specific adjustments
        if language == "java":
            # Java-specific normalization
            normalized_buggy = normalized_buggy.replace('cout', 'System.out.println')
            normalized_fix = normalized_fix.replace('cout', 'System.out.println')
        elif language == "cpp":
            # C++ specific normalization
            normalized_buggy = normalized_buggy.replace('print(', 'cout <<')
            normalized_fix = normalized_fix.replace('print(', 'cout <<')
        
        # Check if buggy line exists (should NOT exist)
        buggy_present = normalized_buggy in normalized_code
        
        # Check if fixed line exists (should exist)
        fix_present = normalized_fix in normalized_code
        
        # Evaluation logic
        if not buggy_present and fix_present:
            return {
                "bug_id": bug.bug_id,
                "status": EvaluationStatus.PASSED.value,
                "message": f"Bug {bug.bug_id} fixed correctly",
                "points_earned": bug.points,
                "max_points": bug.points,
                "buggy_line_found": False,
                "fix_line_found": True,
                "description": bug.description
            }
        elif buggy_present and fix_present:
            return {
                "bug_id": bug.bug_id,
                "status": EvaluationStatus.FAILED.value,
                "message": f"Bug {bug.bug_id} still present alongside fix",
                "points_earned": 0,
                "max_points": bug.points,
                "buggy_line_found": True,
                "fix_line_found": True,
                "description": bug.description
            }
        elif buggy_present and not fix_present:
            return {
                "bug_id": bug.bug_id,
                "status": EvaluationStatus.FAILED.value,
                "message": f"Bug {bug.bug_id} still present, fix not found",
                "points_earned": 0,
                "max_points": bug.points,
                "buggy_line_found": True,
                "fix_line_found": False,
                "description": bug.description
            }
        elif not buggy_present and not fix_present:
            return {
                "bug_id": bug.bug_id,
                "status": EvaluationStatus.FAILED.value,
                "message": f"Bug {bug.bug_id} removed but fix not found",
                "points_earned": 0,
                "max_points": bug.points,
                "buggy_line_found": False,
                "fix_line_found": False,
                "description": bug.description
            }
        
        return {
            "bug_id": bug.bug_id,
            "status": EvaluationStatus.FAILED.value,
            "message": f"Bug {bug.bug_id} evaluation failed",
            "points_earned": 0,
            "max_points": bug.points,
            "buggy_line_found": False,
            "fix_line_found": False,
            "description": bug.description
        }
    
    def evaluate_submission(self, user_code: str, user_language: str) -> Dict:
        """Evaluate user submission against problem requirements"""
        start_time = time.time()
        
        # Language validation
        if user_language not in self.problem.languages:
            return {
                "status": EvaluationStatus.SYSTEM_ERROR.value,
                "score": 0,
                "total_points": self.problem.total_points,
                "bugs_evaluated": [],
                "message": f"Unsupported language: {user_language}",
                "execution_time": 0
            }
        
        # Anti-cheating check
        is_cheating, violations = self.detect_cheating(user_code, user_language)
        if is_cheating:
            return {
                "status": EvaluationStatus.CHEATING_DETECTED.value,
                "score": 0,
                "total_points": self.problem.total_points,
                "bugs_evaluated": [],
                "message": f"Cheating detected: {'; '.join(violations)}",
                "execution_time": 0
            }
        
        # Normalize user code
        normalized_code = self.normalize_code(user_code, user_language)
        
        # Evaluate each bug
        bugs_evaluated = []
        total_score = 0
        
        for bug in self.problem.bugs:
            bug_result = self.evaluate_single_bug(normalized_code, bug, user_language)
            bugs_evaluated.append(bug_result)
            
            if bug_result["status"] == EvaluationStatus.PASSED.value:
                total_score += bug.points
        
        # Determine overall status
        all_passed = all(bug["status"] == EvaluationStatus.PASSED.value for bug in bugs_evaluated)
        any_passed = any(bug["status"] == EvaluationStatus.PASSED.value for bug in bugs_evaluated)
        
        if all_passed:
            overall_status = EvaluationStatus.PASSED
            message = f"All {len(self.problem.bugs)} bugs fixed correctly! Score: {total_score}/{self.problem.total_points}"
        elif any_passed:
            overall_status = EvaluationStatus.FAILED
            passed_count = sum(1 for bug in bugs_evaluated if bug["status"] == EvaluationStatus.PASSED.value)
            message = f"Partial fix: {passed_count}/{len(self.problem.bugs)} bugs fixed. Score: {total_score}/{self.problem.total_points}"
        else:
            overall_status = EvaluationStatus.FAILED
            message = f"No bugs fixed. Score: {total_score}/{self.problem.total_points}"
        
        execution_time = time.time() - start_time
        
        return {
            "status": overall_status.value,
            "score": total_score,
            "total_points": self.problem.total_points,
            "bugs_evaluated": bugs_evaluated,
            "message": message,
            "execution_time": execution_time,
            "fixed_bugs": sum(1 for bug in bugs_evaluated if bug["status"] == EvaluationStatus.PASSED.value),
            "total_bugs": len(self.problem.bugs)
        }
    
    def get_problem_definition(self) -> Dict:
        """Get problem definition for frontend"""
        return {
            "problem_id": self.problem.problem_id,
            "title": self.problem.title,
            "description": self.problem.description,
            "total_points": self.problem.total_points,
            "bugs": [
                {
                    "bug_id": bug.bug_id,
                    "buggy_line": bug.buggy_line,
                    "expected_fix": bug.expected_fix,
                    "description": bug.description,
                    "points": bug.points
                }
                for bug in self.problem.bugs
            ],
            "languages": {
                lang: {
                    "buggy_code": code.buggy_code,
                    "correct_code": code.correct_code
                }
                for lang, code in self.problem.languages.items()
            }
        }

def main():
    """Test the set bits debugging judge"""
    judge = SetBitsDebuggingJudge()
    
    print("=== FixIt Platform - Set Bits Counting Debugging Problem ===\n")
    
    # Test submissions
    test_submissions = [
        {
            "language": "python",
            "code": '''n = int(input())

count = 0

while n > 0:
    if (n & 1) == 1:
        count = count + 1
    n = n >> 1

print(count)''',
            "expected": "PASSED"
        },
        {
            "language": "python", 
            "code": '''n = int(input())

count = 2

while n < 0:
    if (n | 1) = 1:
        count = count + 1
    n = n << 1

print(cout)''',
            "expected": "FAILED"
        },
        {
            "language": "java",
            "code": '''import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int count = 0;
        while (n > 0) {
            if ((n & 1) == 1) {
                count = count + 1;
            }
            n = n >> 1;
        }
        System.out.println(count);
    }
}''',
            "expected": "PASSED"
        },
        {
            "language": "cpp",
            "code": '''#include <iostream>
using namespace std;

int main() {
    int n;
    cin >> n;
    int count = 0;
    while (n > 0) {
        if ((n & 1) == 1) {
            count = count + 1;
        }
        n = n >> 1;
    }
    cout << count;
    return 0;
}''',
            "expected": "PASSED"
        }
    ]
    
    for i, submission in enumerate(test_submissions):
        print(f"{'='*60}")
        print(f"Test Submission {i+1} ({submission['language'].upper()})")
        print(f"{'='*60}")
        
        result = judge.evaluate_submission(submission["code"], submission["language"])
        
        print(f"Status: {result['status']}")
        print(f"Expected: {submission['expected']}")
        print(f"Score: {result['score']}/{result['total_points']}")
        print(f"Fixed Bugs: {result.get('fixed_bugs', 'N/A')}/{result.get('total_bugs', 'N/A')}")
        print(f"Message: {result['message']}")
        print(f"Execution Time: {result['execution_time']:.4f}s")
        
        print("\nBug Evaluation Details:")
        for bug_result in result["bugs_evaluated"]:
            print(f"  {bug_result['bug_id']}: {bug_result['status']}")
            print(f"    {bug_result['message']}")
            print(f"    Points: {bug_result['points_earned']}/{bug_result['max_points']}")
        
        print()

if __name__ == "__main__":
    main()

File: set-bits-debugging-judge.js
// FixIt Platform - Set Bits Counting Problem Integration (Node.js)
// Multi-language debugging problem with line comparison evaluation

const express = require('express');

class SetBitsDebuggingJudge {
    constructor() {
        this.problem = this.createSetBitsProblem();
    }

    createSetBitsProblem() {
        // Bug definitions (common across all languages)
        const bugs = [
            {
                bugId: "count_initialization",
                buggyLine: "count = 2",
                expectedFix: "count = 0",
                description: "Count should start from 0",
                points: 10
            },
            {
                bugId: "loop_condition",
                buggyLine: "while n < 0",
                expectedFix: "while n > 0",
                description: "Loop should run while n is positive",
                points: 15
            },
            {
                bugId: "bitwise_operator",
                buggyLine: "if (n | 1)",
                expectedFix: "if (n & 1)",
                description: "Use AND to check the last bit",
                points: 15
            },
            {
                bugId: "comparison_operator",
                buggyLine: "= 1",
                expectedFix: "== 1",
                description: "Use comparison operator instead of assignment",
                points: 15
            },
            {
                bugId: "shift_direction",
                buggyLine: "n = n << 1",
                expectedFix: "n = n >> 1",
                description: "Shift right to remove processed bit",
                points: 15
            },
            {
                bugId: "print_typo",
                buggyLine: "print(cout)",
                expectedFix: "print(count)",
                description: "Variable name typo",
                points: 10
            }
        ];

        // Language-specific code
        const languages = {
            python: {
                buggyCode: `n = int(input())

count = 2

while n < 0:
    if (n | 1) = 1:
        count = count + 1
    n = n << 1

print(cout)`,
                correctCode: `n = int(input())

count = 0

while n > 0:
    if (n & 1) == 1:
        count = count + 1
    n = n >> 1

print(count)`
            },
            java: {
                buggyCode: `import java.util.*;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        int count = 2;

        while (n < 0) {
            if ((n | 1) = 1) {
                count = count + 1;
            }
            n = n << 1;
        }

        System.out.println(cout);
    }
}`,
                correctCode: `import java.util.*;

public class Main {
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();

        int count = 0;

        while (n > 0) {
            if ((n & 1) == 1) {
                count = count + 1;
            }
            n = n >> 1;
        }

        System.out.println(count);
    }
}`
            },
            cpp: {
                buggyCode: `#include <iostream>
using namespace std;

int main() {

    int n;
    cin >> n;

    int count = 2;

    while (n < 0) {
        if ((n | 1) = 1) {
            count = count + 1;
        }
        n = n << 1;
    }

    cout << cout;

    return 0;
}`,
                correctCode: `#include <iostream>
using namespace std;

int main() {

    int n;
    cin >> n;

    int count = 0;

    while (n > 0) {
        if ((n & 1) == 1) {
            count = count + 1;
        }
        n = n >> 1;
    }

    cout << count;

    return 0;
}`
            }
        };

        return {
            problemId: 1,
            title: "Count Set Bits - Debug the Algorithm",
            description: "Given an integer N, count the number of set bits (1s) in its binary representation. Fix all bugs in the provided code.",
            bugs,
            totalPoints: 80,
            languages
        };
    }

    normalizeCode(code, language) {
        if (!code) return "";
        
        // Convert to string and strip whitespace
        code = String(code).trim();
        
        // Normalize line endings
        code = code.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
        
        // Split into lines and normalize each line
        const lines = code.split('\n');
        const normalizedLines = [];
        
        for (let line of lines) {
            // Remove leading/trailing whitespace
            line = line.trim();
            
            // Normalize internal whitespace (multiple spaces -> single space)
            line = line.replace(/\s+/g, ' ');
            
            // Language-specific comment removal
            if (language === "python") {
                line = line.split('#')[0];  // Remove Python comments
            } else if (language === "java" || language === "cpp") {
                line = line.split('//')[0];  // Remove C++/Java comments
                // Remove C-style comments
                if (line.includes('/*')) {
                    line = line.split('/*')[0];
                }
            }
            
            // Remove trailing whitespace again
            line = line.trim();
            
            // Only add non-empty lines
            if (line) {
                normalizedLines.push(line);
            }
        }
        
        return normalizedLines.join('\n');
    }

    normalizeLine(line) {
        if (!line) return "";
        
        // Convert to string and strip whitespace
        line = String(line).trim();
        
        // Normalize internal whitespace
        line = line.replace(/\s+/g, ' ');
        
        return line;
    }

    detectCheating(code, language) {
        const violations = [];
        const normalizedCode = this.normalizeCode(code, language);
        
        // Check for hardcoded outputs (only obvious patterns)
        const hardcodedPatterns = [
            /print\s*\(\s*\d+\s*\)/gi,           // print(42)
            /cout\s*<<\s*\d+/gi,                  // cout << 42
            /System\.out\.println\s*\(\s*\d+\s*\)/gi,  // System.out.println(42)
            /return\s+\d+/gi,                       // return 42
        ];
        
        for (const pattern of hardcodedPatterns) {
            const matches = normalizedCode.match(pattern);
            if (matches) {
                violations.push(`Hardcoded output detected: ${pattern}`);
            }
        }
        
        // Check for suspicious comments (only if they contain actual code)
        const lines = normalizedCode.split('\n');
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            
            // Only flag if line has actual comment patterns with code keywords
            if ((('#' in line && /if|while|for|return|print/i.test(line)) ||
                ('//' in line && /if|while|for|return|print/i.test(line)))) {
                violations.push(`Suspicious commented code at line ${i + 1}: ${line}`);
            }
        }
        
        return { isCheating: violations.length > 0, violations };
    }

    evaluateSingleBug(normalizedCode, bug, language) {
        // Normalize both buggy line and expected fix
        const normalizedBuggy = this.normalizeLine(bug.buggyLine);
        const normalizedFix = this.normalizeLine(bug.expectedFix);
        
        // Language-specific adjustments
        let adjustedBuggy = normalizedBuggy;
        let adjustedFix = normalizedFix;
        
        if (language === "java") {
            // Java-specific normalization
            adjustedBuggy = adjustedBuggy.replace('print(', 'System.out.println(');
            adjustedFix = adjustedFix.replace('print(', 'System.out.println(');
            adjustedBuggy = adjustedBuggy.replace('cout', 'System.out.println');
            adjustedFix = adjustedFix.replace('cout', 'System.out.println');
        } else if (language === "cpp") {
            // C++ specific normalization
            adjustedBuggy = adjustedBuggy.replace('print(', 'cout <<');
            adjustedFix = adjustedFix.replace('print(', 'cout <<');
        }
        
        // Check if buggy line exists (should NOT exist)
        const buggyPresent = normalizedCode.includes(adjustedBuggy);
        
        // Check if fixed line exists (should exist)
        const fixPresent = normalizedCode.includes(adjustedFix);
        
        // Evaluation logic
        if (!buggyPresent && fixPresent) {
            return {
                bugId: bug.bugId,
                status: "PASSED",
                message: `Bug ${bug.bugId} fixed correctly`,
                pointsEarned: bug.points,
                maxPoints: bug.points,
                buggyLineFound: false,
                fixLineFound: true,
                description: bug.description
            };
        } else if (buggyPresent && fixPresent) {
            return {
                bugId: bug.bugId,
                status: "FAILED",
                message: `Bug ${bug.bugId} still present alongside fix`,
                pointsEarned: 0,
                maxPoints: bug.points,
                buggyLineFound: true,
                fixLineFound: true,
                description: bug.description
            };
        } else if (buggyPresent && !fixPresent) {
            return {
                bugId: bug.bugId,
                status: "FAILED",
                message: `Bug ${bug.bugId} still present, fix not found`,
                pointsEarned: 0,
                maxPoints: bug.points,
                buggyLineFound: true,
                fixLineFound: false,
                description: bug.description
            };
        } else if (!buggyPresent && !fixPresent) {
            return {
                bugId: bug.bugId,
                status: "FAILED",
                message: `Bug ${bug.bugId} removed but fix not found`,
                pointsEarned: 0,
                maxPoints: bug.points,
                buggyLineFound: false,
                fixLineFound: false,
                description: bug.description
            };
        }
        
        return {
            bugId: bug.bugId,
            status: "FAILED",
            message: `Bug ${bug.bugId} evaluation failed`,
            pointsEarned: 0,
            maxPoints: bug.points,
            buggyLineFound: false,
            fixLineFound: false,
            description: bug.description
        };
    }

    evaluateSubmission(userCode, userLanguage) {
        const startTime = Date.now();
        
        // Language validation
        if (!this.problem.languages[userLanguage]) {
            return {
                status: "SYSTEM_ERROR",
                score: 0,
                totalPoints: this.problem.totalPoints,
                bugsEvaluated: [],
                message: `Unsupported language: ${userLanguage}`,
                executionTime: 0
            };
        }
        
        // Anti-cheating check
        const { isCheating, violations } = this.detectCheating(userCode, userLanguage);
        if (isCheating) {
            return {
                status: "CHEATING_DETECTED",
                score: 0,
                totalPoints: this.problem.totalPoints,
                bugsEvaluated: [],
                message: `Cheating detected: ${violations.join('; ')}`,
                executionTime: 0
            };
        }
        
        // Normalize user code
        const normalizedCode = this.normalizeCode(userCode, userLanguage);
        
        // Evaluate each bug
        const bugsEvaluated = [];
        let totalScore = 0;
        
        for (const bug of this.problem.bugs) {
            const bugResult = this.evaluateSingleBug(normalizedCode, bug, userLanguage);
            bugsEvaluated.push(bugResult);
            
            if (bugResult.status === "PASSED") {
                totalScore += bug.points;
            }
        }
        
        // Determine overall status
        const allPassed = bugsEvaluated.every(bug => bug.status === "PASSED");
        const anyPassed = bugsEvaluated.some(bug => bug.status === "PASSED");
        
        let overallStatus, message;
        if (allPassed) {
            overallStatus = "PASSED";
            message = `All ${this.problem.bugs.length} bugs fixed correctly! Score: ${totalScore}/${this.problem.totalPoints}`;
        } else if (anyPassed) {
            overallStatus = "FAILED";
            const passedCount = bugsEvaluated.filter(bug => bug.status === "PASSED").length;
            message = `Partial fix: ${passedCount}/${this.problem.bugs.length} bugs fixed. Score: ${totalScore}/${this.problem.totalPoints}`;
        } else {
            overallStatus = "FAILED";
            message = `No bugs fixed. Score: ${totalScore}/${this.problem.totalPoints}`;
        }
        
        const executionTime = Date.now() - startTime;
        
        return {
            status: overallStatus,
            score: totalScore,
            totalPoints: this.problem.totalPoints,
            bugsEvaluated,
            message,
            executionTime,
            fixedBugs: bugsEvaluated.filter(bug => bug.status === "PASSED").length,
            totalBugs: this.problem.bugs.length
        };
    }

    getProblemDefinition() {
        return {
            problemId: this.problem.problemId,
            title: this.problem.title,
            description: this.problem.description,
            totalPoints: this.problem.totalPoints,
            bugs: this.problem.bugs.map(bug => ({
                bugId: bug.bugId,
                buggyLine: bug.buggyLine,
                expectedFix: bug.expectedFix,
                description: bug.description,
                points: bug.points
            })),
            languages: Object.keys(this.problem.languages).reduce((acc, lang) => {
                acc[lang] = {
                    buggyCode: this.problem.languages[lang].buggyCode,
                    correctCode: this.problem.languages[lang].correctCode
                };
                return acc;
            }, {})
        };
    }
}

// Test the judge
function testSetBitsJudge() {
    console.log('=== FixIt Platform - Set Bits Counting Debugging Problem ===\n');
    
    const judge = new SetBitsDebuggingJudge();
    
    // Test submissions
    const testSubmissions = [
        {
            language: "python",
            code: `n = int(input())

count = 0

while n > 0:
    if (n & 1) == 1:
        count = count + 1
    n = n >> 1

print(count)`,
            expected: "PASSED"
        },
        {
            language: "python", 
            code: `n = int(input())

count = 2

while n < 0:
    if (n | 1) = 1:
        count = count + 1
    n = n << 1

print(cout)`,
            expected: "FAILED"
        },
        {
            language: "java",
            code: `import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int count = 0;
        while (n > 0) {
            if ((n & 1) == 1) {
                count = count + 1;
            }
            n = n >> 1;
        }
        System.out.println(count);
    }
}`,
            expected: "PASSED"
        },
        {
            language: "cpp",
            code: `#include <iostream>
using namespace std;

int main() {
    int n;
    cin >> n;
    int count = 0;
    while (n > 0) {
        if ((n & 1) == 1) {
            count = count + 1;
        }
        n = n >> 1;
    }
    cout << count;
    return 0;
}`,
            expected: "PASSED"
        }
    ];
    
    for (let i = 0; i < testSubmissions.length; i++) {
        const submission = testSubmissions[i];
        console.log('='.repeat(60));
        console.log(`Test Submission ${i + 1} (${submission.language.toUpperCase()})`);
        console.log('='.repeat(60));
        
        const result = judge.evaluateSubmission(submission.code, submission.language);
        
        console.log(`Status: ${result.status}`);
        console.log(`Expected: ${submission.expected}`);
        console.log(`Score: ${result.score}/${result.totalPoints}`);
        console.log(`Fixed Bugs: ${result.fixedBugs}/${result.totalBugs}`);
        console.log(`Message: ${result.message}`);
        console.log(`Execution Time: ${result.executionTime}ms`);
        
        console.log('\nBug Evaluation Details:');
        for (const bugResult of result.bugsEvaluated) {
            console.log(`  ${bugResult.bugId}: ${bugResult.status}`);
            console.log(`    ${bugResult.message}`);
            console.log(`    Points: ${bugResult.pointsEarned}/${bugResult.maxPoints}`);
        }
        
        console.log();
    }
    
    // Display problem definition
    console.log('='.repeat(60));
    console.log('PROBLEM DEFINITION');
    console.log('='.repeat(60));
    
    const problemDef = judge.getProblemDefinition();
    console.log(`Problem ID: ${problemDef.problemId}`);
    console.log(`Title: ${problemDef.title}`);
    console.log(`Description: ${problemDef.description}`);
    console.log(`Total Points: ${problemDef.totalPoints}`);
    console.log(`Languages: ${Object.keys(problemDef.languages).join(', ')}`);
    
    console.log('\nBug Definitions:');
    for (const bug of problemDef.bugs) {
        console.log(`  ${bug.bugId}:`);
        console.log(`    Buggy: ${bug.buggyLine}`);
        console.log(`    Fixed: ${bug.expectedFix}`);
        console.log(`    Points: ${bug.points}`);
        console.log(`    Description: ${bug.description}`);
    }
}

// Run tests
testSetBitsJudge();

module.exports = { SetBitsDebuggingJudge };
